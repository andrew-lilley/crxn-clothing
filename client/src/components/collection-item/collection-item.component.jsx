import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { selectActiveCurrencyLangCode } from '../../redux/currency/currency.selectors';
import { applyFxRate } from '../../redux/currency/currency.util'; 
import { addItem } from '../../redux/cart/cart.actions';
import Price from '../../components/price/price.component';
import { AssetLocation } from '../../utils';
import {
  CollectionItemContainer,
  CollectionFooterContainer,
  AddButton,
  BackgroundImage,
  NameContainer,
  PriceContainer
} from './collection-item.styles';

export const CollectionItem = ({ item, addItem, langCode }) => {

  const { name, price, imageUrl } = item;

  const sale_price = applyFxRate(price, langCode);

  return (
    <CollectionItemContainer>
      <BackgroundImage
        style={{
          backgroundImage: `url(${AssetLocation(imageUrl)})`
        }}
      />
      <CollectionFooterContainer>
        <NameContainer>{name}</NameContainer>
        <PriceContainer><Price price={sale_price} /></PriceContainer>
      </CollectionFooterContainer>
      <AddButton onClick={() => addItem(item, langCode)} inverted>Add to cart</AddButton>
    </CollectionItemContainer>
  )
}

const mapStateToProps = createStructuredSelector({
  langCode: selectActiveCurrencyLangCode
});

const mapDispatchToProps = dispatch => ({
  addItem: (item, langCode) => dispatch(addItem(item, langCode))
});

export default connect(mapStateToProps, mapDispatchToProps)(CollectionItem);