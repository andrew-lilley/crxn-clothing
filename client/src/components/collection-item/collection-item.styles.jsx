import styled from 'styled-components';
import CustomButton from '../custom-button/custom-button.component';

export const CollectionItemContainer = styled.div`
  width: 22vw;
  display: flex;
  flex-direction: column;
  height: 350px;
  align-items: center;
  position: relative;
  &:hover {
    > div:first-child {
      opacity: 0.8;
    }
    button {
      opacity: 0.85;
      display: flex;
    }
  }

  @media screen and (max-width: 800px) {
    width: 40vw;

    &:hover {
      > div:first-child {
        opacity: unset;
      }
      button {
        opacity: unset;
      }
    }
  }
`;

export const AddButton = styled(CustomButton)`
  width: 80%;
  opacity: 0.7;
  position: absolute;
  top: 255px;
  display: none;

  @media screen and (max-width: 800px) {
    display: block;
    opacity: 0.9;
    min-width: unset;
    padding: 0 10px;
  }
`;

// This is required for testing.
AddButton.displayName = 'AddButton';

export const BackgroundImage = styled.div`
  width: 100%;
  height: 95%;
  background-size: cover;
  background-position: center;
  margin-bottom: 5px;
  background-image: ${({ imageUrl }) => `url(${imageUrl})`};
`;

// This is required for testing.
BackgroundImage.displayName = 'BackgroundImage';

export const CollectionFooterContainer = styled.div`
  width: 100%;
  height: 5%;
  display: flex;
  justify-content: space-between;
  font-size: 18px;
`;

// This is required for testing.
CollectionFooterContainer.displayName = 'CollectionFooterContainer';

export const NameContainer = styled.span`
  margin-bottom: 15px;
`;

// This is required for testing.
NameContainer.displayName = 'NameContainer';

export const PriceContainer = styled.span`
  text-align: right;
`;

// This is required for testing.
PriceContainer.displayName = 'PriceContainer';