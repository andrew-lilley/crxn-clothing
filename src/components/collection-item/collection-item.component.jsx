import React from 'react';
import { connect } from 'react-redux';
import { addItem } from '../../redux/cart/cart.actions';
import Price from '../../components/price/price.component';
import {
  CollectionItemContainer,
  CollectionFooterContainer,
  AddButton,
  BackgroundImage,
  NameContainer,
  PriceContainer
} from './collection-item.styles';

export const CollectionItem = ({ item, addItem }) => {

  const { name, price, imageUrl } = item;

  return (
    <CollectionItemContainer>
      <BackgroundImage
        style={{
          backgroundImage: `url(${imageUrl})`
        }}
      />
      <CollectionFooterContainer>
        <NameContainer>{name}</NameContainer>
        <PriceContainer><Price price={price} /></PriceContainer>
      </CollectionFooterContainer>
      <AddButton onClick={() => addItem(item)} inverted>Add to cart</AddButton>
    </CollectionItemContainer>
  )
}

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(addItem(item))
});

export default connect(null, mapDispatchToProps)(CollectionItem);