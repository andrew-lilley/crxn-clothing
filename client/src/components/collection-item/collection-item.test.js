import React from 'react';
import { shallow } from 'enzyme';
import { CollectionItem } from './collection-item.component';

describe('CollectionItem component', () => {
  let wrapper;
  let mockAddItem;
  const imageUrl = 'images/shop-img/sneakers/adidas-nmd.png';
  const mockName = 'Adidas NMD';
  const mockPrice = 220;

  beforeEach(() => {
    mockAddItem = jest.fn();

    const mockProps = {
      item: {
        imageUrl: imageUrl,
        price: mockPrice,
        name: mockName
      },
      addItem: mockAddItem
    };

    wrapper = shallow(<CollectionItem { ...mockProps } />);
  });

  test('should render CollectionItem component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call addItem when AddButton clicked', () => {
    wrapper.find('AddButton').simulate('click');

    expect(mockAddItem).toHaveBeenCalled();
  });

});