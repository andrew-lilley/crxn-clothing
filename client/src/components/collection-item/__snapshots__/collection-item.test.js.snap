// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionItem component should render CollectionItem component 1`] = `
<styled.div>
  <BackgroundImage
    style={
      Object {
        "backgroundImage": "url(adidas-nmd.png)",
      }
    }
  />
  <CollectionFooterContainer>
    <NameContainer>
      Adidas NMD
    </NameContainer>
    <PriceContainer>
      <Connect(Price)
        price={220}
      />
    </PriceContainer>
  </CollectionFooterContainer>
  <AddButton
    inverted={true}
    onClick={[Function]}
  >
    Add to cart
  </AddButton>
</styled.div>
`;
